src/main/java/org/quiltmc/loader/api/QuiltLoader.java

```diff
@@ -0,0 +1,191 @@
+	 *
+	 * <p>The entrypoint is declared in the {@code fabric.mod.json} as following:
+	 * <pre><blockquote>
+	 *   "entrypoints": {
+	 *     "&lt;a key&gt;": [
+	 *       &lt;a list of entrypoint declarations&gt;
+	 *     ]
+	 *   }
+	 * </blockquote></pre>
+	 * Multiple keys can be present in the {@code entrypoints} section.</p>
+	 *
+	 * <p>An entrypoint declaration indicates that an arbitrary notation is sent
+	 * to a {@link LanguageAdapter} to offer an instance of entrypoint. It is
+	 * either a string, or an object. An object declaration
+	 * is of this form:<pre><blockquote>
+	 *   {
+	 *     "adapter": "&lt;a custom adatper&gt;"
+	 *     "value": "&lt;an arbitrary notation&gt;"
+	 *   }
+	 * </blockquote></pre>
+	 * A string declaration {@code <an arbitrary notation>} is equivalent to
+	 * <pre><blockquote>
+	 *   {
+	 *     "adapter": "default"
+	 *     "value": "&lt;an arbitrary notation&gt;"
+	 *   }
+	 * </blockquote></pre>
+	 * where the {@code default} adapter is the {@linkplain LanguageAdapter adapter}
+	 * offered by Fabric Loader. </p>
+	 *
+	 * @param key  the key in entrypoint declaration in {@code fabric.mod.json}
+	 * @param type the type of entrypoints
+	 * @param <T>  the type of entrypoints
+	 * @return the entrypoint containers related to this key
+	 * @throws EntrypointException if a problem arises during entrypoint creation
+	 * @see LanguageAdapter
+	 */
+	<T> List<EntrypointContainer<T>> getEntrypointContainers(String key, Class<T> type);
+
+	/**
+	 * Get the current mapping resolver.
+	 *
+	 * <p>When performing reflection, a mod should always query the mapping resolver for
+	 * the remapped names of members than relying on other heuristics.</p>
+	 *
+	 * @return the current mapping resolver instance
+	 * @since 0.4.1
+	 */
+	MappingResolver getMappingResolver();
+
+	/**
+	 * Gets the container for a given mod.
+	 *
+	 * @param id the ID of the mod
+	 * @return the mod container, if present
+	 */
+	Optional<ModContainer> getModContainer(String id);
+
+	/**
+	 * Gets all mod containers.
+	 *
+	 * @return a collection of all loaded mod containers
+	 */
+	Collection<ModContainer> getAllMods();
+
+	/**
+	 * Checks if a mod with a given ID is loaded.
+	 *
+	 * @param id the ID of the mod, as defined in {@code fabric.mod.json}
+	 * @return whether or not the mod is present in this Fabric Loader instance
+	 */
+	boolean isModLoaded(String id);
+
+	/**
+	 * Checks if Fabric Loader is currently running in a "development"
+	 * environment. Can be used for enabling debug mode or additional checks.
+	 *
+	 * <p>This should not be used to make assumptions on certain features,
+	 * such as mappings, but as a toggle for certain functionalities.</p>
+	 *
+	 * @return whether or not Loader is currently in a "development"
+	 * environment
+	 */
+	boolean isDevelopmentEnvironment();
+
+	/**
+	 * Get the current environment type.
+	 *
+	 * @return the current environment type
+	 */
+	EnvType getEnvironmentType();
```

Comment:
We do need to think about game specific environment details and how to represent those without the loader having a hard requirement on the Minecraft game implementations.

Reply:
What if EnvType was some form of interface? Not sure exactly how that would work, but it would allow other games to specify their own environments

Reply:
Sadly `EnvType` is used as an annotation value for stripping - so we'd need to rethink how that is applied if we want to change it.

Reply:
We could introduce a MinecraftQuiltLoader class which exposes these game specific attributes as an option?

Reply:
That makes more sense to me than trying to generalise this between games. Technically that could be part of a "minecraft loader plugin", but we can at least make the `MinecraftQuiltLoader` class now?

Reply:
Yeah that seems fine for now